File: src/main/java/me/narutopig/bors/EnchantmentWrapper.java
package me.narutopig.bors;

import org.bukkit.NamespacedKey;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.enchantments.EnchantmentTarget;
import org.bukkit.inventory.ItemStack;

import java.util.Arrays;

import static me.narutopig.bors.Util.power;

public class EnchantmentWrapper extends Enchantment {
    private final String name;
    private final int maxLevel;
    private Enchantment[] conflicts = new Enchantment[] {};
    public ItemStack cost;

    public EnchantmentWrapper(String key, String name, int maxLevel, ItemStack cost, Enchantment[] conflicts) {
        super(NamespacedKey.minecraft(key));
        this.name = name;
        this.maxLevel = maxLevel;
        this.cost = cost;
        this.conflicts = conflicts;
    }

    public EnchantmentWrapper(String key, String name, int maxLevel, ItemStack cost) {
        super(NamespacedKey.minecraft(key));
        this.name = name;
        this.maxLevel = maxLevel;
        this.cost = cost;
    }

    public ItemStack getCost() {
        return cost;
    }

    public ItemStack getCost(int level) {
        int newLevel = power(2, level - 1);

        ItemStack newCost = cost.clone();

        newCost.setAmount(newLevel);

        return newCost;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public int getMaxLevel() {
        return maxLevel;
    }

    @Override
    public int getStartLevel() {
        return 1;
    }

    @Override
    public EnchantmentTarget getItemTarget() {
        return EnchantmentTarget.BREAKABLE;
    }

    @Override
    public boolean isTreasure() {
        return false;
    }

    @Override
    public boolean isCursed() {
        return false;
    }

    @Override
    public boolean conflictsWith(Enchantment other) {
        return Arrays.stream(conflicts).filter(enchantment -> enchantment.getKey().equals(other.getKey())).toList().size() > 0;
    }

    @Override
    public boolean canEnchantItem(ItemStack item) {
        return true;
    }
}

File: src/main/kotlin/io/github/narutopig/bors/EnchantmentWrapper.kt
package io.github.narutopig.bors

import io.github.narutopig.bors.util.General
import org.bukkit.NamespacedKey
import org.bukkit.enchantments.Enchantment
import org.bukkit.enchantments.EnchantmentTarget
import org.bukkit.inventory.ItemStack
import java.util.*
import javax.annotation.Nonnull

class EnchantmentWrapper : Enchantment {
    private val name: String
    private val maxLevel: Int
    private var conflicts: Array<Enchantment> = arrayOf()
    private var cost: ItemStack

    constructor(key: String, name: String, maxLevel: Int, cost: ItemStack, conflicts: Array<Enchantment>) : super(
        NamespacedKey.minecraft(
            key
        )
    ) {
        this.name = name
        this.maxLevel = maxLevel
        this.cost = cost
        this.conflicts = conflicts
    }

    constructor(key: String, name: String, maxLevel: Int, cost: ItemStack) : super(
        NamespacedKey.minecraft(
            key
        )
    ) {
        this.name = name
        this.maxLevel = maxLevel
        this.cost = cost
    }

    fun getCost(level: Int): ItemStack {
        val newLevel = General.power(2, level - 1) * this.cost.amount
        val newCost = cost.clone()
        newCost.amount = newLevel
        return newCost
    }

    @Nonnull
    override fun getName(): String {
        return name
    }

    override fun getMaxLevel(): Int {
        return maxLevel
    }

    override fun getStartLevel(): Int {
        return 1
    }

    @Nonnull
    override fun getItemTarget(): EnchantmentTarget {
        return EnchantmentTarget.BREAKABLE
    }

    override fun isTreasure(): Boolean {
        return false
    }

    override fun isCursed(): Boolean {
        return false
    }

    override fun conflictsWith(@Nonnull other: Enchantment): Boolean {
        return Arrays.stream(conflicts).anyMatch { enchantment: Enchantment -> enchantment.key == other.key }
    }

    override fun canEnchantItem(@Nonnull item: ItemStack): Boolean {
        return true
    }
}